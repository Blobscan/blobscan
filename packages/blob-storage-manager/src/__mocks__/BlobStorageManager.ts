import { beforeEach } from "vitest";
import { mockDeep, mockReset } from "vitest-mock-extended";

import type { BlobStorageManager } from "../BlobStorageManager";

beforeEach(() => {
  mockReset(blobStorageManager);
});

const blobStorageManager = mockDeep<BlobStorageManager>();
export default blobStorageManager;
