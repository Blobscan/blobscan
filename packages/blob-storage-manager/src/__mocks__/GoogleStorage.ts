import type { Storage } from "@google-cloud/storage";
import type { StorageOptions } from "@google-cloud/storage";
import { vi } from "vitest";

import { FILE_URI } from "../../test/fixtures";
import { GoogleStorage } from "../storages";
import type {
  GoogleCredentials,
  GoogleStorageConfig,
} from "../storages/GoogleStorage";

vi.mock("@google-cloud/storage", async () => {
  return {
    Storage: vi.fn().mockImplementation((config: StorageOptions) => {
      return {
        projectId: config.projectId,
        apiEndpoint: config.apiEndpoint,
        getServiceAccount: vi.fn().mockReturnValue([
          {
            emailAddress: config.credentials
              ? (config.credentials as GoogleCredentials).client_email
              : undefined,
          },
        ]),
        getBuckets: vi.fn().mockResolvedValue([[{ name: "mock-bucket" }]]),
        bucket: vi.fn().mockReturnValue({
          file: vi.fn().mockImplementation((uri: string) => {
            return {
              download: vi.fn().mockImplementation(() => {
                if (uri === FILE_URI) {
                  return Buffer.from("mock-data");
                }
                throw new Error("File not found");
              }),
              save: vi.fn((data: string) => data),
            };
          }),
        }),
        createBucket: vi.fn().mockResolvedValue(undefined),
      };
    }),
  };
});

export class GoogleStorageMock extends GoogleStorage {
  constructor(config: GoogleStorageConfig) {
    super(config);
  }

  get bucketName(): string {
    return this._bucketName;
  }

  get storageClient(): Storage {
    return this._storageClient;
  }
}
