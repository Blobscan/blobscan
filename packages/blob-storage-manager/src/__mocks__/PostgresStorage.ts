import type { PrismaClient } from "@blobscan/db";
import prisma from "@blobscan/db/prisma/__mocks__/client";

import { BLOB_HASH, RAW_DATA } from "../../test/fixtures";
import { PostgresStorage } from "../storages";

export class PostgresStorageMock extends PostgresStorage {
  constructor() {
    super();

    this.client = prisma as unknown as PrismaClient;
  }

  getBlob(versionedHash: string): Promise<string> {
    // mock prisma query
    prisma.blobData.findFirstOrThrow.mockResolvedValue({
      id: BLOB_HASH,
      data: RAW_DATA,
    });

    return super.getBlob(versionedHash);
  }

  async storeBlob(
    _: number,
    versionedHash: string,
    blobData: string
  ): Promise<string> {
    // mock prisma query
    prisma.blobData.upsert.mockResolvedValue({
      id: BLOB_HASH,
      data: RAW_DATA,
    });

    return super.storeBlob(_, versionedHash, blobData);
  }
}
