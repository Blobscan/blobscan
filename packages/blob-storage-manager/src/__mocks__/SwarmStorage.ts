import { Bee, BeeDebug } from "@ethersphere/bee-js";
import { vi } from "vitest";

import { SWARM_REFERENCE } from "../../test/fixtures";
import type { SwarmClient, SwarmStorageConfig } from "../storages";
import { SwarmStorage } from "../storages";

vi.mock("@ethersphere/bee-js", async () => {
  return {
    Bee: vi.fn().mockImplementation((endpoint) => {
      return {
        url: endpoint,
        checkConnection: vi
          .fn()
          .mockResolvedValueOnce({
            status: "ok",
          })
          .mockRejectedValueOnce(new Error("Bee is not healthy: not ok"))
          .mockResolvedValueOnce({
            status: "ok",
          }),
        downloadFile: vi.fn().mockImplementation((reference) => {
          if (reference === SWARM_REFERENCE) {
            return {
              data: {
                text() {
                  return "mock-data";
                },
              },
            };
          }
          throw new Error("File not found");
        }),
        uploadFile: vi.fn().mockResolvedValue({
          reference: "mock-reference",
        }),
      };
    }),
    BeeDebug: vi.fn().mockImplementation((endpoint) => {
      return {
        url: endpoint,
        getHealth: vi
          .fn()
          .mockResolvedValueOnce({
            status: "ok",
          })
          .mockResolvedValueOnce({
            status: "ok",
          })
          .mockResolvedValueOnce({
            status: "not ok",
          }),
        getAllPostageBatch: vi
          .fn()
          .mockResolvedValueOnce([
            {
              batchID: "mock-batch-id",
            },
          ])
          .mockResolvedValueOnce([])
          .mockResolvedValueOnce([]),
      };
    }),
  };
});

export class SwarmStorageMock extends SwarmStorage {
  constructor({ beeDebugEndpoint, beeEndpoint, chainId }: SwarmStorageConfig) {
    super({ beeDebugEndpoint, beeEndpoint, chainId });

    this._swarmClient = {
      bee: new Bee(beeEndpoint),
      beeDebug: beeDebugEndpoint ? new BeeDebug(beeDebugEndpoint) : undefined,
    };
  }

  get swarmClient(): SwarmClient {
    return this._swarmClient;
  }
}
