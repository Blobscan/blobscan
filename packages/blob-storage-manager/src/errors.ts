class ErrorException extends Error {
  constructor(message: string, cause?: Error | Error[]) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class BlobStorageError extends ErrorException {
  constructor(storageName: string, message: string, cause?: Error) {
    super(`${storageName} failed: ${message}`, cause);
  }
}

export class StorageCreationError extends BlobStorageError {
  cause: BlobStorageError;

  constructor(storageName: string, message: string, cause: BlobStorageError) {
    super(storageName, `Creation failed: ${message}`);

    this.cause = cause;
  }
}

export class BlobStorageManagerError extends ErrorException {
  constructor(message: string, cause?: Error | BlobStorageError[]) {
    // const storageErrorsMessage = storageErrors?.length
    //   ? `Failed storages: ${storageErrors.map((e) => e.message).join(", ")}`
    //   : "";
    super(message, cause);
  }
}
