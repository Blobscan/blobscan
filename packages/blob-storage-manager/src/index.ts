export { BlobStorageManager } from "./BlobStorageManager";
export * from "./blob-storage-manager";
export { env } from "./env";
export type { Environment } from "./env";
export * from "./errors";
export { GoogleStorage, PostgresStorage, SwarmStorage } from "./storages";
export * from "./types";
