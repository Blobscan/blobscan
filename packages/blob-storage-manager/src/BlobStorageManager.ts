import { api, SemanticAttributes } from "@blobscan/open-telemetry";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManagerError } from "./errors";
import type { BlobStorageError } from "./errors";
import { tracer, updateBlobStorageMetrics } from "./instrumentation";
import type {
  Blob,
  BlobReference,
  BlobStorageName,
  StorageOf,
  StoreOptions,
} from "./types";
import { calculateBlobBytes, removeDuplicatedStorages } from "./utils";

export class BlobStorageManager {
  #blobStorages: BlobStorage[];

  constructor(blobStorages: BlobStorage[]) {
    if (!blobStorages.length) {
      throw new BlobStorageManagerError("No blob storages provided");
    }

    this.#blobStorages = removeDuplicatedStorages(blobStorages);
  }

  getStorage<T extends BlobStorageName>(name: T): StorageOf<T> | undefined {
    const blobStorage = this.#blobStorages.find(
      (storage): storage is StorageOf<T> => storage.name === name
    );

    return blobStorage as StorageOf<T> | undefined;
  }

  hasStorage<T extends BlobStorageName>(name: T): boolean {
    return !!this.getStorage(name);
  }

  async getBlob(
    ...blobReferences: BlobReference<BlobStorageName>[]
  ): Promise<{ data: string; storage: BlobStorageName } | null> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "getBlob",
        },
      },
      async (span) => {
        const availableReferences = blobReferences.filter(({ storage }) =>
          this.hasStorage(storage)
        );

        try {
          const blob = await Promise.any(
            availableReferences.map(({ reference, storage: storageName }) =>
              tracer.startActiveSpan(
                "blob_storage_manager:storage",
                {
                  attributes: {
                    storage: storageName,
                    [SemanticAttributes.CODE_FUNCTION]: "getBlob",
                  },
                },
                async (storageSpan) => {
                  const start = performance.now();
                  const storage = this.getStorage(storageName);

                  if (!storage) {
                    throw new BlobStorageManagerError(
                      `Storage "${storageName}" is not available`
                    );
                  }

                  const blob = await storage
                    .getBlob(reference)
                    .then((data) => ({
                      data,
                      storage: storageName,
                    }));
                  const end = performance.now();

                  storageSpan.end();

                  updateBlobStorageMetrics({
                    storage: storageName,
                    blobSize: calculateBlobBytes(blob.data),
                    direction: "received",
                    duration: end - start,
                  });

                  return blob;
                }
              )
            )
          );

          span.end();

          return blob;
        } catch (e) {
          const message = "Failed to get blob from any of the storages";
          let err: BlobStorageManagerError;

          // Aggregate errors are thrown when all the promises fail
          if (e instanceof AggregateError) {
            err = new BlobStorageManagerError(
              message,
              e.errors as BlobStorageError[]
            );
          } else {
            err = new BlobStorageManagerError(message, e as Error);
          }

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);

          throw err;
        } finally {
          span.end();
        }
      }
    );
  }

  async storeBlob(
    { data, versionedHash }: Blob,
    opts?: StoreOptions
  ): Promise<{
    references: BlobReference<BlobStorageName>[];
    errors: BlobStorageError[];
  }> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
        },
      },
      async (span) => {
        const selectedStorages = this.#getSelectedStorages(opts);

        const results = await Promise.allSettled(
          selectedStorages.map((storage) => {
            return tracer.startActiveSpan(
              "blob_storage_manager:storage",
              {
                attributes: {
                  storage: storage.name,
                  [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
                },
              },
              async (storageSpan) => {
                const start = performance.now();
                const blobReference = await storage
                  .storeBlob(versionedHash, data)
                  .then<BlobReference<BlobStorageName>>((reference) => ({
                    reference,
                    storage: storage.name,
                  }));
                const end = performance.now();

                storageSpan.end();

                updateBlobStorageMetrics({
                  blobSize: calculateBlobBytes(data),
                  direction: "sent",
                  duration: end - start,
                  storage: storage.name,
                });

                return blobReference;
              }
            );
          })
        );
        const references = results
          .filter(
            (
              res
            ): res is PromiseFulfilledResult<BlobReference<BlobStorageName>> =>
              res.status === "fulfilled"
          )
          .map((res) => res.value);
        const storageErrors = results
          .filter((r) => r.status === "rejected")
          .map((r) => (r as PromiseRejectedResult).reason as BlobStorageError);

        if (!references.length) {
          const err = new BlobStorageManagerError(
            `Failed to upload blob with hash "${versionedHash}" to any storage`,
            storageErrors
          );

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);
          throw err;
        }

        span.end();

        return {
          references,
          errors: storageErrors,
        };
      }
    );
  }

  #getSelectedStorages(
    { selectedStorages }: StoreOptions = { selectedStorages: [] }
  ): BlobStorage[] {
    const uniqueInputStorageNames = Array.from(new Set(selectedStorages));
    const selectedAvailableStorages = uniqueInputStorageNames.length
      ? this.#blobStorages.filter((s) =>
          uniqueInputStorageNames.includes(s.name)
        )
      : // Use all available storages if none are selected
        this.#blobStorages;

    if (
      uniqueInputStorageNames.length &&
      selectedAvailableStorages.length !== uniqueInputStorageNames.length
    ) {
      const selectdAvailableStorageNames = selectedAvailableStorages.map(
        (s) => s.name
      );
      const missingStorageNames = uniqueInputStorageNames.filter(
        (storageName) => !selectdAvailableStorageNames.includes(storageName)
      );

      throw new BlobStorageManagerError(
        `Some of the selected storages are not available: ${missingStorageNames.join(
          ", "
        )}"`
      );
    }

    return selectedAvailableStorages;
  }
}
