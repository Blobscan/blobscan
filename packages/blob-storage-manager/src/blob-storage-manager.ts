import { BlobStorageManager } from "./BlobStorageManager";
import { env } from "./env";
import { GoogleStorage, PostgresStorage, SwarmStorage } from "./storages";

let blobStorageManager: BlobStorageManager | undefined;

export async function getBlobStorageManager(): Promise<BlobStorageManager> {
  if (!blobStorageManager) {
    const blobStorages = [];
    const [googleStorage, googleStorageError] =
      await GoogleStorage.tryCreateFromEnv(env);
    const [postgresStorage, postgresStorageError] =
      await PostgresStorage.tryCreateFromEnv(env);
    const [swarmStorage, swarmStorageError] =
      await SwarmStorage.tryCreateFromEnv(env);

    if (googleStorage) {
      blobStorages.push(googleStorage);
    }

    if (postgresStorage) {
      blobStorages.push(postgresStorage);
    }

    if (swarmStorage) {
      blobStorages.push(swarmStorage);
    }

    blobStorageManager = new BlobStorageManager(blobStorages, env.CHAIN_ID);
  }

  return blobStorageManager;
}
