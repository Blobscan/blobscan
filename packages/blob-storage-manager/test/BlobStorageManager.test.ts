import { beforeAll, describe, expect, it } from "vitest";
import type { DeepMockProxy } from "vitest-mock-extended";
import { mockDeep } from "vitest-mock-extended";

import { testValidError } from "@blobscan/test";

import { GoogleStorage, PostgresStorage, env } from "../src";
import { BlobStorageManager } from "../src/BlobStorageManager";
import { SwarmStorageMock as SwarmStorage } from "../src/__mocks__";
import { BlobStorageError, BlobStorageManagerError } from "../src/errors";
import type { BlobStorageName } from "../src/types";
import { BLOB_DATA, BLOB_HASH, FILE_URI, SWARM_REFERENCE } from "./fixtures";

if (!env.BEE_ENDPOINT) {
  throw new Error("BEE_ENDPOINT test env var is not set");
}

const BEE_ENDPOINT = env.BEE_ENDPOINT;

describe("BlobStorageManager", () => {
  let blobStorageManager: BlobStorageManager;
  let postgresStorage: PostgresStorage;
  let googleStorage: GoogleStorage;
  let swarmStorage: SwarmStorage;

  let failingPostgresStorage: DeepMockProxy<PostgresStorage>;
  let failingGoogleStorage: DeepMockProxy<GoogleStorage>;
  let failingSwarmStorage: DeepMockProxy<SwarmStorage>;

  beforeAll(() => {
    if (!env.GOOGLE_STORAGE_BUCKET_NAME) {
      throw new BlobStorageError(
        "GoogleStorage",
        "GOOGLE_STORAGE_BUCKET_NAME is not set"
      );
    }

    postgresStorage = new PostgresStorage({ chainId: env.CHAIN_ID });
    googleStorage = new GoogleStorage({
      chainId: env.CHAIN_ID,
      bucketName: env.GOOGLE_STORAGE_BUCKET_NAME,
      apiEndpoint: env.GOOGLE_STORAGE_API_ENDPOINT,
      projectId: env.GOOGLE_STORAGE_PROJECT_ID,
    });
    swarmStorage = new SwarmStorage({
      chainId: env.CHAIN_ID,
      beeEndpoint: BEE_ENDPOINT,
      beeDebugEndpoint: env.BEE_DEBUG_ENDPOINT,
    });

    blobStorageManager = new BlobStorageManager([
      postgresStorage,
      googleStorage,
      swarmStorage,
    ]);

    failingPostgresStorage = mockDeep<PostgresStorage>();
    failingGoogleStorage = mockDeep<GoogleStorage>();
    failingSwarmStorage = mockDeep<SwarmStorage>();

    failingPostgresStorage.storeBlob.mockRejectedValue(
      new BlobStorageError(
        "PostgresStorage",
        "Failed to upload blob to postgres"
      )
    );
    failingGoogleStorage.storeBlob.mockRejectedValue(
      new BlobStorageError("GoogleStorage", "Failed to upload blob to google")
    );
    failingSwarmStorage.storeBlob.mockRejectedValue(
      new BlobStorageError("SwarmStorage", "Failed to upload blob to swarm")
    );
  });

  describe("constructor", () => {
    it("should throw an error if no blob storages are provided", () => {
      expect(() => new BlobStorageManager([])).toThrow(
        "No blob storages provided"
      );
    });
  });

  describe("getStorage", () => {
    it("should return the correct blob storage for a given name", async () => {
      expect(blobStorageManager.getStorage("POSTGRES")).toEqual(
        postgresStorage
      );

      expect(blobStorageManager.getStorage("GOOGLE")).toEqual(googleStorage);
      expect(blobStorageManager.getStorage("SWARM")).toEqual(swarmStorage);
    });
  });

  describe("getBlob", async () => {
    it("should return the blob data and storage name", async () => {
      const result = await blobStorageManager.getBlob(
        {
          reference: BLOB_HASH,
          storage: "POSTGRES",
        },
        {
          reference: FILE_URI,
          storage: "GOOGLE",
        },
        {
          reference: SWARM_REFERENCE,
          storage: "SWARM",
        }
      );

      expect([
        { data: "0x6d6f636b2d64617461", storage: "POSTGRES" },
        { data: "mock-data", storage: "GOOGLE" },
        { data: "mock-data", storage: "SWARM" },
      ]).toContainEqual(result);
    });

    testValidError(
      "should throw an error if the blob storage is not found",
      async () => {
        const UNKNOWN_BLOB_HASH = "0x6d6f636b2d64617461";
        const UNKNOWN_FILE_URI = "1/6d/6f/636b2d64617461.txt";
        const UNKNOWN_SWARM_REFERENCE = "123456789abcdef";

        await blobStorageManager.getBlob(
          {
            reference: UNKNOWN_BLOB_HASH,
            storage: "POSTGRES",
          },
          {
            reference: UNKNOWN_FILE_URI,
            storage: "GOOGLE",
          },
          {
            reference: UNKNOWN_SWARM_REFERENCE,
            storage: "SWARM",
          }
        );
      },
      BlobStorageManagerError,
      {
        checkCause: true,
      }
    );
  });

  describe("storeBlob", () => {
    const blob = { data: BLOB_DATA, versionedHash: BLOB_HASH };
    it("should store the blob in all available storages", async () => {
      const result = await blobStorageManager.storeBlob(blob);

      expect(result.references).matchSnapshot();
      expect(result.errors).toMatchSnapshot();
    });

    it("should store a blob in a specific storage if provided", async () => {
      const selectedStorage: BlobStorageName = "POSTGRES";

      const result = await blobStorageManager.storeBlob(blob, {
        selectedStorages: [selectedStorage],
      });

      const blobReference = result.references[0];

      expect(
        result.references.length,
        "Returned blob storage refs length mismatch"
      ).toBe(1);
      expect(blobReference?.reference, "Blob storage ref mismatch").toBe(
        BLOB_HASH
      );
      expect(blobReference?.storage, "Blob storage mismatch").toBe(
        selectedStorage
      );
    });

    testValidError(
      "should throw an error when one of the selected blob storages wasn't found",
      async () => {
        const selectedStorages: BlobStorageName[] = ["POSTGRES", "GOOGLE"];
        const singleStorageBSM = new BlobStorageManager([swarmStorage]);

        await singleStorageBSM.storeBlob(blob, {
          selectedStorages: selectedStorages,
        });
      },
      BlobStorageManagerError
    );

    it("should return errors for failed uploads", async () => {
      const newHash = "0x6d6f636b2d64617461";
      const blob = { data: "New data", versionedHash: newHash };

      const blobStorageManager = new BlobStorageManager([
        failingPostgresStorage,
        googleStorage,
        failingSwarmStorage,
      ]);

      const result = await blobStorageManager.storeBlob(blob);

      expect(result).toMatchSnapshot();
    });

    testValidError(
      "should throw an error if all uploads fail",
      async () => {
        const newBlobStorageManager = new BlobStorageManager([
          failingPostgresStorage,
          failingGoogleStorage,
          failingSwarmStorage,
        ]);

        const blob = {
          data: "New data",
          versionedHash: "0x6d6f636b2d64617461",
        };

        await newBlobStorageManager.storeBlob(blob);
      },
      BlobStorageManagerError,
      {
        checkCause: true,
      }
    );
  });
});
