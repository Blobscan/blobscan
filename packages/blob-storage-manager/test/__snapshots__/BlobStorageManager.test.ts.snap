// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlobStorageManager > getBlob > should throw an error if the blob storage is not found 1`] = `"Failed to get blob from any of the storages"`;

exports[`BlobStorageManager > getBlob > should throw an error if the blob storage is not found 2`] = `
[
  [BlobStorageError: PostgresStorage failed: Failed to get blob with uri "0x6d6f636b2d64617461"],
  [BlobStorageError: GoogleStorage failed: Failed to get blob with uri "1/6d/6f/636b2d64617461.txt"],
  [BlobStorageError: SwarmStorageMock failed: Failed to get blob with uri "123456789abcdef"],
]
`;

exports[`BlobStorageManager > storeBlob > should return errors for failed uploads 1`] = `
{
  "errors": [
    [BlobStorageError: PostgresStorage failed: Failed to upload blob to postgres],
    [BlobStorageError: SwarmStorage failed: Failed to upload blob to swarm],
  ],
  "references": [
    {
      "reference": "70118930558/6d/6f/63/6d6f636b2d64617461.txt",
      "storage": "GOOGLE",
    },
  ],
}
`;

exports[`BlobStorageManager > storeBlob > should store the blob in all available storages 1`] = `
[
  {
    "reference": "0x0100eac880c712dba4346c88ab564fa1b79024106f78f732cca49d8a68e4c174",
    "storage": "POSTGRES",
  },
  {
    "reference": "70118930558/01/00/ea/0100eac880c712dba4346c88ab564fa1b79024106f78f732cca49d8a68e4c174.txt",
    "storage": "GOOGLE",
  },
  {
    "reference": "mock-reference",
    "storage": "SWARM",
  },
]
`;

exports[`BlobStorageManager > storeBlob > should store the blob in all available storages 2`] = `[]`;

exports[`BlobStorageManager > storeBlob > should throw an error if all uploads fail 1`] = `"Failed to upload blob with hash \\"0x6d6f636b2d64617461\\" to any storage"`;

exports[`BlobStorageManager > storeBlob > should throw an error if all uploads fail 2`] = `
[
  [BlobStorageError: PostgresStorage failed: Failed to upload blob to postgres],
  [BlobStorageError: GoogleStorage failed: Failed to upload blob to google],
  [BlobStorageError: SwarmStorage failed: Failed to upload blob to swarm],
]
`;

exports[`BlobStorageManager > storeBlob > should throw an error when one of the selected blob storages wasn't found 1`] = `"Some of the selected storages are not available: POSTGRES, GOOGLE\\""`;
